cmd/relay-gossip/node/relaynode.go

```diff
@@ -0,0 +1,201 @@
+	"google.golang.org/grpc/credentials"
+)
+
+// GossipRelayConfig configures a gossip relay node.
+type GossipRelayConfig struct {
+	Network         string
+	PeerWith        []string
+	Addr            string
+	DataDir         string
+	IdentityPath    string
+	CertPath        string
+	Insecure        bool
+	DrandPublicGRPC string
+}
+
+// GossipRelayNode is a gossip relay runtime.
+type GossipRelayNode struct {
+	bootstrap []ma.Multiaddr
+	ds        *bds.Datastore
+	priv      crypto.PrivKey
+	h         host.Host
+	ps        *pubsub.PubSub
+	t         *pubsub.Topic
+	opts      []grpc.DialOption
+	addr      []ma.Multiaddr
+	done      chan struct{}
+}
+
+// NewGossipRelayNode starts a new gossip relay node.
+func NewGossipRelayNode(cfg *GossipRelayConfig) (*GossipRelayNode, error) {
+	bootstrap, err := ParseMultiaddrSlice(cfg.PeerWith)
+	if err != nil {
+		return nil, xerrors.Errorf("parsing peer-with: %w", err)
+	}
+
+	ds, err := bds.NewDatastore(cfg.DataDir, nil)
+	if err != nil {
+		return nil, xerrors.Errorf("opening datastore: %w", err)
+	}
+
+	priv, err := lp2p.LoadOrCreatePrivKey(cfg.IdentityPath)
+	if err != nil {
+		return nil, xerrors.Errorf("loading p2p key: %w", err)
+	}
+
+	h, ps, err := lp2p.ConstructHost(ds, priv, cfg.Addr, bootstrap)
+	if err != nil {
+		return nil, xerrors.Errorf("constructing host: %w", err)
+	}
+
+	addr, err := h.Network().InterfaceListenAddresses()
+	if err != nil {
+		return nil, xerrors.Errorf("getting InterfaceListenAddresses: %w", err)
+	}
+
+	for _, a := range addr {
+		log.Infof("%s/p2p/%s\n", a, h.ID())
+	}
+
+	t, err := ps.Join(lp2p.PubSubTopic(cfg.Network))
+	if err != nil {
+		return nil, xerrors.Errorf("joining topic: %w", err)
+	}
+
+	opts := []grpc.DialOption{}
+	if cfg.CertPath != "" {
+		creds, err := credentials.NewClientTLSFromFile(cfg.CertPath, "")
+		if err != nil {
+			return nil, xerrors.Errorf("loading cert file: %w", err)
+		}
+		opts = append(opts, grpc.WithTransportCredentials(creds))
+	} else if cfg.Insecure {
+		opts = append(opts, grpc.WithInsecure())
+	} else {
+		opts = append(opts, grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{})))
+	}
+	g := &GossipRelayNode{
+		bootstrap: bootstrap,
+		ds:        ds,
+		priv:      priv,
+		h:         h,
+		ps:        ps,
+		t:         t,
+		opts:      opts,
+		addr:      addr,
+		done:      make(chan struct{}),
+	}
+	go g.start(cfg.DrandPublicGRPC)
+	return g, nil
+}
+
+// Addr returns the gossipsub multiaddresses of this relay node.
+func (g *GossipRelayNode) Addr() []ma.Multiaddr {
+	b := make([]ma.Multiaddr, len(g.addr))
+	for i, a := range g.addr {
+		m, err := ma.NewMultiaddr(fmt.Sprintf("%s/p2p/%s", a, g.h.ID()))
+		if err != nil {
+			panic(err)
+		}
+		b[i] = m
+	}
+	return b
+}
+
+// Shutdown stops the relay node.
+func (g *GossipRelayNode) Shutdown() {
+	close(g.done)
+}
+
+func ParseMultiaddrSlice(peer []string) ([]ma.Multiaddr, error) {
+	out := make([]ma.Multiaddr, len(peer))
+	for i, peer := range peer {
```

Comment:
To make this a little more readable by removing the singular and shadowing:

```suggestion
func ParseMultiaddrSlice(peers []string) ([]ma.Multiaddr, error) {
	out := make([]ma.Multiaddr, len(peers))
	for i, peer := range peers {
```